### PR TITLE
Fix #1650: Make sure credentials for STS get caller identity uses the configuration parameters

### LINKED ISSUE #1650: remote_state backend S3 bucket parameters not set when using an assumed role profile
# Overview
When using an AWS Profile configured to assume an IAM Role to run terragrunt, the bucket is created successfully but none of the bucket parameters are set.  This includes properties like Taggin, SSE, Versioning, and Bucket Policy.  

**NOTE: Initially I thought this issue was isolated to AWS SSO Profiles but I have been able to reproduce this issue with _any_ assumed role AWS Profile!**

The following error is observed during the bucket creation/init sequence...
```
 Create S3 bucket with retry {REDACTED_BUCKET_NAME} returned an error: NoCredentialProviders: no valid providers in chain. Deprecated.
```
- Terragrunt Version: 0.28.22
- Terraform Version: 0.14.10

# Steps to Reproduce
### Show Identity
**NOTE: The behavior is the same whether `AWS_PROFILE` and/or `AWS_DEFAULT_PROFILE` are set to the desired profile**
```
❯ echo $AWS_PROFILE
terragrunt_remote_state_s3_test
```
```
❯ aws sts get-caller-identity
{
    "UserId": "AROASUTQMXEREDACTED00:botocore-session-1619019040",
    "Account": "{REDACTED}",
    "Arn": "arn:aws:sts::{REDACTED:assumed-role/CloudAdmin/botocore-session-1619019040"
}
```
### Terragrunt.hcl
```hcl
terraform {
  source = ".//test_module"
}

remote_state {
  generate = {
    path      = "backend.tf"
    if_exists = "overwrite_terragrunt"
  }
  backend = "s3"
  config = {
    encrypt = true
    bucket  = "terragrunt-backend-s3-test003"
    key     = "terraform.tfstate"
    region  = "us-west-2"

    s3_bucket_tags = {
        Owner       = "foo"
        Environment = "test"
    }
  }
}
```
### test_module/test.tf
```hcl
# Does nothing but generate an output based on caller identity
provider "aws" {
  region = "us-west-2"
}

data "aws_caller_identity" "test" {}

output "caller_id" {
  value = data.aws_caller_identity.test
}
```
### Terragrunt Apply (first-run with backend S3 bucket creation)
```
❯ terragrunt apply --terragrunt-log-level debug
DEBU[0000] Did not find any locals block: skipping evaluation.
DEBU[0000] Running command: terraform --version          prefix=[/tmp/tgtest]
DEBU[0001] Terraform version: 0.14.10
DEBU[0001] Reading Terragrunt config file at /tmp/tgtest/terragrunt.hcl
DEBU[0001] Did not find any locals block: skipping evaluation.
DEBU[0001] Downloading Terraform configurations from file:///tmp/tgtest into /tmp/tgtest/.terragrunt-cache/I4xXOEwr4vXM7HkWXn2JJ0W772M/lJw6GWXuZS6lhlJTWNpiqodZCtY
DEBU[0001] Copying files from /tmp/tgtest into /tmp/tgtest/.terragrunt-cache/I4xXOEwr4vXM7HkWXn2JJ0W772M/lJw6GWXuZS6lhlJTWNpiqodZCtY/test_module
DEBU[0001] Setting working directory to /tmp/tgtest/.terragrunt-cache/I4xXOEwr4vXM7HkWXn2JJ0W772M/lJw6GWXuZS6lhlJTWNpiqodZCtY/test_module
DEBU[0001] Generated file /tmp/tgtest/.terragrunt-cache/I4xXOEwr4vXM7HkWXn2JJ0W772M/lJw6GWXuZS6lhlJTWNpiqodZCtY/test_module/backend.tf.  prefix=[/tmp/tgtest]
DEBU[0001] Initializing remote state for the s3 backend  prefix=[/tmp/tgtest]
Remote state S3 bucket terragrunt-backend-s3-test003 does not exist or you don't have permissions to access it. Would you like Terragrunt to create it? (y/n) y
DEBU[0007] Create S3 bucket with retry terragrunt-backend-s3-test003  prefix=[/tmp/tgtest]
DEBU[0007] Creating S3 bucket terragrunt-backend-s3-test003  prefix=[/tmp/tgtest]
DEBU[0008] Created S3 bucket terragrunt-backend-s3-test003  prefix=[/tmp/tgtest]
DEBU[0008] Waiting for bucket terragrunt-backend-s3-test003 to be created  prefix=[/tmp/tgtest]
DEBU[0008] S3 bucket terragrunt-backend-s3-test003 created.  prefix=[/tmp/tgtest]
DEBU[0008] Enabling root access to S3 bucket terragrunt-backend-s3-test003  prefix=[/tmp/tgtest]
ERRO[0013] Create S3 bucket with retry terragrunt-backend-s3-test003 returned an error: NoCredentialProviders: no valid providers in chain. Deprecated.
	For verbose messaging see aws.Config.CredentialsChainVerboseErrors. Sleeping for 10s and will try again.  prefix=[/tmp/tgtest]
DEBU[0023] Create S3 bucket with retry terragrunt-backend-s3-test003  prefix=[/tmp/tgtest]
DEBU[0023] Creating S3 bucket terragrunt-backend-s3-test003  prefix=[/tmp/tgtest]
DEBU[0023] Looks like you're already creating bucket terragrunt-backend-s3-test003 at the same time. Will not attempt to create it again.  prefix=[/tmp/tgtest]
DEBU[0023] Waiting for bucket terragrunt-backend-s3-test003 to be created  prefix=[/tmp/tgtest]
DEBU[0024] S3 bucket terragrunt-backend-s3-test003 created.  prefix=[/tmp/tgtest]
WARN[0024] Versioning is not enabled for the remote state S3 bucket terragrunt-backend-s3-test003. We recommend enabling versioning so that you can roll back to previous versions of your Terraform state in case of error.  prefix=[/tmp/tgtest]
DEBU[0024] Running command: terraform init               prefix=[/tmp/tgtest]
```
```
❯ cat terragrunt-debug.tfvars.json
{}%
```
```
❯ aws s3api get-bucket-tagging --bucket terragrunt-backend-s3-test003

An error occurred (NoSuchTagSet) when calling the GetBucketTagging operation: The TagSet does not exist
```
```
❯ aws s3api get-bucket-versioning --bucket terragrunt-backend-s3-test003
```
```
❯ aws s3api get-bucket-encryption --bucket terragrunt-backend-s3-test003

An error occurred (ServerSideEncryptionConfigurationNotFoundError) when calling the GetBucketEncryption operation: The server side encryption configuration was not found
```
```
❯ aws s3api get-bucket-policy --bucket terragrunt-backend-s3-test003

An error occurred (NoSuchBucketPolicy) when calling the GetBucketPolicy operation: The bucket policy does not exist
```
### Terragrunt.hcl
```hcl
terraform {
  source = ".//test_module"
}

remote_state {
  generate = {
    path      = "backend.tf"
    if_exists = "overwrite_terragrunt"
  }
  backend = "s3"
  config = {
    encrypt = true
    bucket  = "terragrunt-backend-s3-test000"
    key     = "terraform.tfstate"
    region  = "us-west-2"
  }
}
```
### test_module/test.tf
```hcl
# Does nothing but generate an output based on caller identity
provider "aws" {
  region = "us-west-2"
}

data "aws_caller_identity" "test" {}

output "caller_id" {
  value = data.aws_caller_identity.test
}
```
### Terragrunt Apply (first-run with backend S3 bucket creation)
```
❯ terragrunt apply --terragrunt-debug --terragrunt-log-level debug
DEBU[0000] Found locals block: evaluating the expressions.
DEBU[0000] Evaluated 2 locals (remaining 0): region, env
DEBU[0000] Running command: terraform --version          prefix=[/tmp/tgtest]
DEBU[0001] Terraform version: 0.14.10
DEBU[0001] Reading Terragrunt config file at /tmp/tgtest/terragrunt.hcl
DEBU[0001] Found locals block: evaluating the expressions.
DEBU[0001] Evaluated 2 locals (remaining 0): region, env
DEBU[0001] Downloading Terraform configurations from file:///tmp/tgtest into /tmp/tgtest/.terragrunt-cache/I4xXOEwr4vXM7HkWXn2JJ0W772M/lJw6GWXuZS6lhlJTWNpiqodZCtY
DEBU[0001] Copying files from /tmp/tgtest into /tmp/tgtest/.terragrunt-cache/I4xXOEwr4vXM7HkWXn2JJ0W772M/lJw6GWXuZS6lhlJTWNpiqodZCtY/mod
DEBU[0001] Setting working directory to /tmp/tgtest/.terragrunt-cache/I4xXOEwr4vXM7HkWXn2JJ0W772M/lJw6GWXuZS6lhlJTWNpiqodZCtY/mod
DEBU[0001] The file path /tmp/tgtest/.terragrunt-cache/I4xXOEwr4vXM7HkWXn2JJ0W772M/lJw6GWXuZS6lhlJTWNpiqodZCtY/mod/backend.tf already exists, but was a previously generated file by terragrunt. Since if_exists for code generation is set to "overwrite_terragrunt", regenerating file.  prefix=[/tmp/tgtest]
DEBU[0001] Generated file /tmp/tgtest/.terragrunt-cache/I4xXOEwr4vXM7HkWXn2JJ0W772M/lJw6GWXuZS6lhlJTWNpiqodZCtY/mod/backend.tf.  prefix=[/tmp/tgtest]
INFO[0001] Debug mode requested: generating debug file terragrunt-debug.tfvars.json in working dir /tmp/tgtest/.terragrunt-cache/I4xXOEwr4vXM7HkWXn2JJ0W772M/lJw6GWXuZS6lhlJTWNpiqodZCtY/mod  prefix=[/tmp/tgtest]
DEBU[0001] The following variables were detected in the terraform module:  prefix=[/tmp/tgtest]
DEBU[0001] []                                            prefix=[/tmp/tgtest]
DEBU[0001] Variables passed to terraform are located in "/tmp/tgtest/terragrunt-debug.tfvars.json"  prefix=[/tmp/tgtest]
DEBU[0001] Run this command to replicate how terraform was invoked:  prefix=[/tmp/tgtest]
DEBU[0001] 	terraform apply -var-file="/tmp/tgtest/terragrunt-debug.tfvars.json" "/tmp/tgtest/.terragrunt-cache/I4xXOEwr4vXM7HkWXn2JJ0W772M/lJw6GWXuZS6lhlJTWNpiqodZCtY/mod"  prefix=[/tmp/tgtest]
DEBU[0001] Initializing remote state for the s3 backend  prefix=[/tmp/tgtest]
Remote state S3 bucket terragrunt-backend-s3-test000 does not exist or you don't have permissions to access it. Would you like Terragrunt to create it? (y/n) y
DEBU[0006] Create S3 bucket with retry terragrunt-backend-s3-test000  prefix=[/tmp/tgtest]
DEBU[0006] Creating S3 bucket terragrunt-backend-s3-test000  prefix=[/tmp/tgtest]
DEBU[0007] Created S3 bucket terragrunt-backend-s3-test000  prefix=[/tmp/tgtest]
DEBU[0007] Waiting for bucket terragrunt-backend-s3-test000 to be created  prefix=[/tmp/tgtest]
DEBU[0007] S3 bucket terragrunt-backend-s3-test000 created.  prefix=[/tmp/tgtest]
DEBU[0007] Enabling root access to S3 bucket terragrunt-backend-s3-test000  prefix=[/tmp/tgtest]
ERRO[0013] Create S3 bucket with retry terragrunt-backend-s3-test000 returned an error: NoCredentialProviders: no valid providers in chain. Deprecated.
	For verbose messaging see aws.Config.CredentialsChainVerboseErrors. Sleeping for 10s and will try again.  prefix=[/tmp/tgtest]
DEBU[0023] Create S3 bucket with retry terragrunt-backend-s3-test000  prefix=[/tmp/tgtest]
DEBU[0023] Creating S3 bucket terragrunt-backend-s3-test000  prefix=[/tmp/tgtest]
DEBU[0024] Looks like you're already creating bucket terragrunt-backend-s3-test000 at the same time. Will not attempt to create it again.  prefix=[/tmp/tgtest]
DEBU[0024] Waiting for bucket terragrunt-backend-s3-test000 to be created  prefix=[/tmp/tgtest]
DEBU[0024] S3 bucket terragrunt-backend-s3-test000 created.  prefix=[/tmp/tgtest]
WARN[0024] Versioning is not enabled for the remote state S3 bucket terragrunt-backend-s3-test000. We recommend enabling versioning so that you can roll back to previous versions of your Terraform state in case of error.  prefix=[/tmp/tgtest]
DEBU[0024] Running command: terraform init               prefix=[/tmp/tgtest]
```
```
❯ cat terragrunt-debug.tfvars.json
{}%
```
```
❯ aws s3api get-bucket-versioning --bucket terragrunt-backend-s3-test000
```
```
❯ aws s3api get-bucket-encryption --bucket terragrunt-backend-s3-test000

An error occurred (ServerSideEncryptionConfigurationNotFoundError) when calling the GetBucketEncryption operation: The server side encryption configuration was not found
```
```
❯ aws s3api get-bucket-policy --bucket terragrunt-backend-s3-test000

An error occurred (NoSuchBucketPolicy) when calling the GetBucketPolicy operation: The bucket policy does not exist
```

# Workarounds
## Using AWS Credential Environment variables
If you use standard AWS credential env vars (i.e. `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, `AWS_SESSION_TOKEN`) instead of using an AWS Profile (i.e. `AWS_PROFILE` and/or `AWS_DEFAULT_PROFILE` env vars) the bucket creation and parameters are set correctly, as expected, without error.  
**NOTE: It does NOT matter where the AWS cred env vars are sourced from (e.g. AWS IAM User, AWS IAM Role assumption,  AWS SSO generated CLI Creds)**
### Terragrunt Apply (first-run with backend S3 bucket creation)
```
❯ terragrunt apply --terragrunt-debug --terragrunt-log-level debug
DEBU[0000] Did not find any locals block: skipping evaluation.
DEBU[0000] Running command: terraform --version          prefix=[/tmp/tgtest]
DEBU[0001] Terraform version: 0.14.10
DEBU[0001] Reading Terragrunt config file at /tmp/tgtest/terragrunt.hcl
DEBU[0001] Did not find any locals block: skipping evaluation.
DEBU[0001] Downloading Terraform configurations from file:///tmp/tgtest into /tmp/tgtest/.terragrunt-cache/I4xXOEwr4vXM7HkWXn2JJ0W772M/lJw6GWXuZS6lhlJTWNpiqodZCtY
DEBU[0001] Copying files from /tmp/tgtest into /tmp/tgtest/.terragrunt-cache/I4xXOEwr4vXM7HkWXn2JJ0W772M/lJw6GWXuZS6lhlJTWNpiqodZCtY/test_module
DEBU[0001] Setting working directory to /tmp/tgtest/.terragrunt-cache/I4xXOEwr4vXM7HkWXn2JJ0W772M/lJw6GWXuZS6lhlJTWNpiqodZCtY/test_module
DEBU[0001] Generated file /tmp/tgtest/.terragrunt-cache/I4xXOEwr4vXM7HkWXn2JJ0W772M/lJw6GWXuZS6lhlJTWNpiqodZCtY/test_module/backend.tf.  prefix=[/tmp/tgtest]
INFO[0001] Debug mode requested: generating debug file terragrunt-debug.tfvars.json in working dir /tmp/tgtest/.terragrunt-cache/I4xXOEwr4vXM7HkWXn2JJ0W772M/lJw6GWXuZS6lhlJTWNpiqodZCtY/test_module  prefix=[/tmp/tgtest]
DEBU[0001] The following variables were detected in the terraform module:  prefix=[/tmp/tgtest]
DEBU[0001] []                                            prefix=[/tmp/tgtest]
DEBU[0001] Variables passed to terraform are located in "/tmp/tgtest/terragrunt-debug.tfvars.json"  prefix=[/tmp/tgtest]
DEBU[0001] Run this command to replicate how terraform was invoked:  prefix=[/tmp/tgtest]
DEBU[0001] 	terraform apply -var-file="/tmp/tgtest/terragrunt-debug.tfvars.json" "/tmp/tgtest/.terragrunt-cache/I4xXOEwr4vXM7HkWXn2JJ0W772M/lJw6GWXuZS6lhlJTWNpiqodZCtY/test_module"  prefix=[/tmp/tgtest]
DEBU[0001] Initializing remote state for the s3 backend  prefix=[/tmp/tgtest]
Remote state S3 bucket terragrunt-backend-s3-test001 does not exist or you don't have permissions to access it. Would you like Terragrunt to create it? (y/n) y
DEBU[0005] Create S3 bucket with retry terragrunt-backend-s3-test001  prefix=[/tmp/tgtest]
DEBU[0005] Creating S3 bucket terragrunt-backend-s3-test001  prefix=[/tmp/tgtest]
DEBU[0006] Created S3 bucket terragrunt-backend-s3-test001  prefix=[/tmp/tgtest]
DEBU[0006] Waiting for bucket terragrunt-backend-s3-test001 to be created  prefix=[/tmp/tgtest]
DEBU[0006] S3 bucket terragrunt-backend-s3-test001 created.  prefix=[/tmp/tgtest]
DEBU[0006] Enabling root access to S3 bucket terragrunt-backend-s3-test001  prefix=[/tmp/tgtest]
DEBU[0007] Enabled root access to bucket terragrunt-backend-s3-test001  prefix=[/tmp/tgtest]
DEBU[0007] Enabling enforced TLS access for S3 bucket terragrunt-backend-s3-test001  prefix=[/tmp/tgtest]
DEBU[0007] Enabled enforced TLS access for bucket terragrunt-backend-s3-test001  prefix=[/tmp/tgtest]
DEBU[0007] Blocking all public access to S3 bucket terragrunt-backend-s3-test001  prefix=[/tmp/tgtest]
DEBU[0008] Blocked all public access to S3 bucket terragrunt-backend-s3-test001  prefix=[/tmp/tgtest]
DEBU[0008] No tags specified for bucket terragrunt-backend-s3-test001.  prefix=[/tmp/tgtest]
DEBU[0008] Enabling versioning on S3 bucket terragrunt-backend-s3-test001  prefix=[/tmp/tgtest]
DEBU[0008] Enabled versioning on S3 bucket terragrunt-backend-s3-test001  prefix=[/tmp/tgtest]
DEBU[0008] Enabling bucket-wide SSE on AWS S3 bucket terragrunt-backend-s3-test001  prefix=[/tmp/tgtest]
DEBU[0008] Enabled bucket-wide SSE on AWS S3 bucket terragrunt-backend-s3-test001  prefix=[/tmp/tgtest]
DEBU[0008] Access Logging is disabled for the remote state AWS S3 bucket terragrunt-backend-s3-test001  prefix=[/tmp/tgtest]
DEBU[0008] Running command: terraform init               prefix=[/tmp/tgtest]
```
```
❯ cat terragrunt-debug.tfvars.json
{}%
```
```
❯ aws s3api get-bucket-versioning --bucket terragrunt-backend-s3-test001
{
    "Status": "Enabled"
}
```
```
❯ aws s3api get-bucket-encryption --bucket terragrunt-backend-s3-test001
{
    "ServerSideEncryptionConfiguration": {
        "Rules": [
            {
                "ApplyServerSideEncryptionByDefault": {
                    "SSEAlgorithm": "aws:kms"
                },
                "BucketKeyEnabled": false
            }
        ]
    }
}
```
```
❯ aws s3api get-bucket-policy --bucket terragrunt-backend-s3-test001
{
    "Policy": "{\"Version\":\"2012-10-17\",\"Statement\":[{\"Sid\":\"AllowTLSRequestsOnly\",\"Effect\":\"Deny\",\"Principal\":\"*\",\"Action\":\"s3:*\",\"Resource\":[\"arn:aws:s3:::terragrunt-backend-s3-test001\",\"arn:aws:s3:::terragrunt-backend-s3-test001/*\"],\"Condition\":{\"Bool\":{\"aws:SecureTransport\":\"false\"}}}]}"
}
```
## Using Terragrunt `iam_role` config property to assume the role instead of an AWS Profile
**NOTE: This method will not work with assumption of SSO IAM Roles as they are _only_ assumable by the AWS SSO service!**
### Terragrunt.hcl
```hcl
iam_role = "arn:aws:iam::{REDACTED}:role/okta-AWS-CloudAdmin"

terraform {
  source = ".//test_module"
}

remote_state {
  generate = {
    path      = "backend.tf"
    if_exists = "overwrite_terragrunt"
  }
  backend = "s3"
  config = {
    encrypt = true
    bucket  = "terragrunt-backend-s3-test005"
    key     = "terraform.tfstate"
    region  = "us-west-2"

    s3_bucket_tags = {
        Owner       = "foo"
        Environment = "test"
    }
  }
}
```
### test_module/test.tf
```hcl
# Does nothing but generate an output based on caller identity
provider "aws" {
  region = "us-west-2"
}

data "aws_caller_identity" "test" {}

output "caller_id" {
  value = data.aws_caller_identity.test
}
```
### Terragrunt Apply (first-run with backend S3 bucket creation)
```
❯ terragrunt apply --terragrunt-log-level debug
DEBU[0000] Did not find any locals block: skipping evaluation.
DEBU[0000] Running command: terraform --version          prefix=[/tmp/tgtest]
DEBU[0001] Terraform version: 0.14.10
DEBU[0001] Reading Terragrunt config file at /tmp/tgtest/terragrunt.hcl
DEBU[0001] Did not find any locals block: skipping evaluation.
DEBU[0001] Assuming IAM role arn:aws:iam::{REDACTED}:role/okta-AWS-CloudAdmin
DEBU[0002] Downloading Terraform configurations from file:///tmp/tgtest into /tmp/tgtest/.terragrunt-cache/I4xXOEwr4vXM7HkWXn2JJ0W772M/lJw6GWXuZS6lhlJTWNpiqodZCtY
DEBU[0002] Copying files from /tmp/tgtest into /tmp/tgtest/.terragrunt-cache/I4xXOEwr4vXM7HkWXn2JJ0W772M/lJw6GWXuZS6lhlJTWNpiqodZCtY/test_module
DEBU[0002] Setting working directory to /tmp/tgtest/.terragrunt-cache/I4xXOEwr4vXM7HkWXn2JJ0W772M/lJw6GWXuZS6lhlJTWNpiqodZCtY/test_module
DEBU[0002] Generated file /tmp/tgtest/.terragrunt-cache/I4xXOEwr4vXM7HkWXn2JJ0W772M/lJw6GWXuZS6lhlJTWNpiqodZCtY/test_module/backend.tf.  prefix=[/tmp/tgtest]
DEBU[0002] Initializing remote state for the s3 backend  prefix=[/tmp/tgtest]
Remote state S3 bucket terragrunt-backend-s3-test005 does not exist or you don't have permissions to access it. Would you like Terragrunt to create it? (y/n) y
DEBU[0006] Create S3 bucket with retry terragrunt-backend-s3-test005  prefix=[/tmp/tgtest]
DEBU[0006] Creating S3 bucket terragrunt-backend-s3-test005  prefix=[/tmp/tgtest]
DEBU[0007] Created S3 bucket terragrunt-backend-s3-test005  prefix=[/tmp/tgtest]
DEBU[0007] Waiting for bucket terragrunt-backend-s3-test005 to be created  prefix=[/tmp/tgtest]
DEBU[0007] S3 bucket terragrunt-backend-s3-test005 created.  prefix=[/tmp/tgtest]
DEBU[0007] Enabling root access to S3 bucket terragrunt-backend-s3-test005  prefix=[/tmp/tgtest]
DEBU[0008] Enabled root access to bucket terragrunt-backend-s3-test005  prefix=[/tmp/tgtest]
DEBU[0008] Enabling enforced TLS access for S3 bucket terragrunt-backend-s3-test005  prefix=[/tmp/tgtest]
DEBU[0008] Enabled enforced TLS access for bucket terragrunt-backend-s3-test005  prefix=[/tmp/tgtest]
DEBU[0008] Blocking all public access to S3 bucket terragrunt-backend-s3-test005  prefix=[/tmp/tgtest]
DEBU[0008] Blocked all public access to S3 bucket terragrunt-backend-s3-test005  prefix=[/tmp/tgtest]
  1 iam_role = "arn:aws:iam::{REDACTED}:role/okta-AWS-CloudAdmin"
DEBU[0008] Tagging S3 bucket with map[Environment:test Owner:foo]  prefix=[/tmp/tgtest]
DEBU[0008] Tagged S3 bucket with map[Environment:test Owner:foo]  prefix=[/tmp/tgtest]
DEBU[0008] Enabling versioning on S3 bucket terragrunt-backend-s3-test005  prefix=[/tmp/tgtest]
DEBU[0009] Enabled versioning on S3 bucket terragrunt-backend-s3-test005  prefix=[/tmp/tgtest]
DEBU[0009] Enabling bucket-wide SSE on AWS S3 bucket terragrunt-backend-s3-test005  prefix=[/tmp/tgtest]
DEBU[0009] Enabled bucket-wide SSE on AWS S3 bucket terragrunt-backend-s3-test005  prefix=[/tmp/tgtest]
DEBU[0009] Access Logging is disabled for the remote state AWS S3 bucket terragrunt-backend-s3-test005  prefix=[/tmp/tgtest]
DEBU[0009] Running command: terraform init               prefix=[/tmp/tgtest]
```
```
❯ aws s3api get-bucket-tagging --bucket terragrunt-backend-s3-test005
{
    "TagSet": [
        {
            "Key": "Environment",
            "Value": "test"
        },
        {
            "Key": "Owner",
            "Value": "foo"
        }
    ]
}
```
```
❯ aws s3api get-bucket-versioning --bucket terragrunt-backend-s3-test005
{
    "Status": "Enabled"
}
```
```
❯ aws s3api get-bucket-policy --bucket terragrunt-backend-s3-test005
{
    "Policy": "{\"Version\":\"2012-10-17\",\"Statement\":[{\"Sid\":\"AllowTLSRequestsOnly\",\"Effect\":\"Deny\",\"Principal\":\"*\",\"Action\":\"s3:*\",\"Resource\":[\"arn:aws:s3:::terragrunt-backend-s3-test005\",\"arn:aws:s3:::terragrunt-backend-s3-test005/*\"],\"Condition\":{\"Bool\":{\"aws:SecureTransport\":\"false\"}}}]}"
}
```
```
❯ aws s3api get-bucket-encryption --bucket terragrunt-backend-s3-test005
{
    "ServerSideEncryptionConfiguration": {
        "Rules": [
            {
                "ApplyServerSideEncryptionByDefault": {
                    "SSEAlgorithm": "aws:kms"
                },
                "BucketKeyEnabled": false
            }
        ]
    }
}
```
Related: #1129, #1537
Possibly related: #1369

### PR DESCRIPTION
**Pull request by yorinasub17**

### CODE REVIEW

File: remote/remote_state_s3.go [hunk 1]
```diff
@@ -560,10 +560,11 @@ func isBucketAlreadyOwnedByYouError(err error) bool {
...
 
 // Add a policy to allow root access to the bucket
-func EnableRootAccesstoS3Bucket(s3Client *s3.S3, config *RemoteStateConfigS3, terragruntOptions *options.TerragruntOptions) error {
-	terragruntOptions.Logger.Debugf("Enabling root access to S3 bucket %s", config.Bucket)
+func EnableRootAccesstoS3Bucket(s3Client *s3.S3, config *ExtendedRemoteStateConfigS3, terragruntOptions *options.TerragruntOptions) error {
+	bucket := config.remoteStateConfigS3.Bucket
+	terragruntOptions.Logger.Debugf("Enabling root access to S3 bucket %s", bucket)
 
-	accountID, err := aws_helper.GetAWSAccountID(terragruntOptions)
+	accountID, err := aws_helper.GetAWSAccountID(config.GetAwsSessionConfig(), terragruntOptions)
```
Comment (yorinasub17) [inline remote/remote_state_s3.go:567]: This is the source of the bug: this call was ignoring all the session parameters defined in the remote state config.

File: aws_helper/config.go [hunk 2]
```diff
@@ -83,10 +83,33 @@ func CreateAwsSession(config *AwsSessionConfig, terragruntOptions *options.Terra
 	} else if terragruntOptions.IamRole != "" {
 		sess.Config.Credentials = stscreds.NewCredentials(sess, terragruntOptions.IamRole, credentialsOptFn)
 	}
+	return sess, nil
+}
+
+// Returns an AWS session object for the given config region (required), profile name (optional), and IAM role to assume
```
Comment (brikis98) [inline aws_helper/config.go:89]: NIT: Are profile name + IAM role passed in via `terragruntOptions`? If so, we should say so explicitly.

File: aws_helper/config.go [hunk 3]
```diff
@@ -83,10 +83,33 @@ func CreateAwsSession(config *AwsSessionConfig, terragruntOptions *options.Terra
...
+	var sess *session.Session
+	var err error
+	if config == nil {
+		sess, err = session.NewSession()
+		if err != nil {
+			return nil, errors.WithStackTrace(err)
+		}
+		if terragruntOptions.IamRole != "" {
+			sess.Config.Credentials = stscreds.NewCredentials(sess, terragruntOptions.IamRole)
+		}
```
Comment (brikis98) [inline aws_helper/config.go:102]: I see the IAM role being used here, but do we need to do something with the profile here? I think both IAM role and profile are handled automatically in `CreateAwsSessionFromConfig` in the `else` block below, but the profile seems to be missing from this `if` block?

Comment (yorinasub17) [reply aws_helper/config.go:102]: I must be missing something obvious. We don't have support for AWS profiles specified via the CLI or config (which is what gets translated to `terragruntOptions`), but instead rely on the SDK environment variable `AWS_PROFILE`, which is handled via `session.NewSession`. Did I miss a struct attribute?

File: aws_helper/config.go (hunk 2 above)
Comment (yorinasub17) [reply aws_helper/config.go]: Updated in 861668f

File: aws_helper/config.go (hunk 3 above)
Comment (yorinasub17) [reply aws_helper/config.go:105]: Ok I think there might have been confusion from the functions docs, so I took a stab at updating it. Does it make more sense what is going on now?

861668f

Comment (brikis98) [reply aws_helper/config.go:105]: Ah, yea, the docs were indicating there was a profile we needed to handle, but with the clarification, it looks like that should already have been baked into the passed in `config`.

**Review by brikis98 — APPROVED** [review]
LGTM

**Comment by yorinasub17** [comment]
Thanks for review! Going to merge this in.

**Comment by stevie-** [comment]
I've also seen issues with `run_cmd(["aws", "...."])` function and AWS SSO Login via `AWS_PROFILE`. Possible that hsi fixes these issues too?

**Pull request merged**